PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h

```diff
@@ -271,3 +272,134 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
     return out;
   }
 };
+
+template <typename T, typename TProd>
+class SimpleFlatTableProducerBaseLumi : public edm::one::EDProducer<edm::EndLuminosityBlockProducer> {
+public:
+  SimpleFlatTableProducerBaseLumi(edm::ParameterSet const &params)
+      : name_(params.getParameter<std::string>("name")),
+        doc_(params.existsAs<std::string>("doc") ? params.getParameter<std::string>("doc") : ""),
+        extension_(params.existsAs<bool>("extension") ? params.getParameter<bool>("extension") : false),
+        skipNonExistingSrc_(
+            params.existsAs<bool>("skipNonExistingSrc") ? params.getParameter<bool>("skipNonExistingSrc") : false),
+        src_(skipNonExistingSrc_ ? mayConsume<TProd>(params.getParameter<edm::InputTag>("src"))
+                                 : consumes<TProd,edm::InLumi>(params.getParameter<edm::InputTag>("src"))) {
+    edm::ParameterSet const &varsPSet = params.getParameter<edm::ParameterSet>("variables");
+    for (const std::string &vname : varsPSet.getParameterNamesForType<edm::ParameterSet>()) {
+      const auto &varPSet = varsPSet.getParameter<edm::ParameterSet>(vname);
+      const std::string &type = varPSet.getParameter<std::string>("type");
+      if (type == "int")
+        vars_.push_back(std::make_unique<IntVar>(vname, varPSet));
+      else if (type == "float")
+        vars_.push_back(std::make_unique<FloatVar>(vname, varPSet));
+      else if (type == "uint8")
+        vars_.push_back(std::make_unique<UInt8Var>(vname, varPSet));
+      else if (type == "bool")
+        vars_.push_back(std::make_unique<BoolVar>(vname, varPSet));
+      else
+        throw cms::Exception("Configuration", "unsupported type " + type + " for variable " + vname);
+    }
+
+    produces<nanoaod::FlatTable,edm::Transition::EndLuminosityBlock>();
+  }
+
+  ~SimpleFlatTableProducerBaseLumi() override {}
+
+  // this is to be overriden by the child class
+  virtual std::unique_ptr<nanoaod::FlatTable> fillTable(const edm::LuminosityBlock &iLumi,
+                                                        const edm::Handle<TProd> &prod) const = 0;
+
+  void produce(edm::Event &iEvent, const edm::EventSetup &iSetup) override {
+    // do nothing 
+  }
+
+  void endLuminosityBlockProduce(edm::LuminosityBlock& iLumi, const edm::EventSetup &iSetup) final {
+    edm::Handle<TProd> src;
+    iLumi.getByToken(src_, src);
+
+    std::unique_ptr<nanoaod::FlatTable> out = fillTable(iLumi, src);
+    out->setDoc(doc_);
+
+    iLumi.put(std::move(out));
+  }
+
+protected:
+  const std::string name_;
+  const std::string doc_;
+  const bool extension_;
+  const bool skipNonExistingSrc_;
+  const edm::EDGetTokenT<TProd> src_;
+
+  class VariableBase {
```

Comment:
This class appears to be independent of `T` or `TProd`. Moving its definition outside of the `SimpleFlatTableProducerBaseLumi` would generate less code (only one definition instead of one per instantiation of `SimpleFlatTableProducerBaseLumi`).

Reply:
Dear @makortel,

As with the previous cases, I based the 'Lumi version' of the module off the existing SimpleFlatTableProducerBase: https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h#L65

Therefore, I think this question is best addressed to the original developers. If it were to be updated, then I think it would make sense that it is updated consistently in both modules by the original developers.

Best,
Mateusz